packages/mask/src/extension/popups/hook/useFriends.ts

```diff
@@ -43,90 +44,94 @@ export const PlatformSort: Record<NextIDPlatform, number> = {
     const currentPersona = useCurrentPersona()
-    return useAsyncRetry(async () => {
-        const values = await Services.Identity.queryRelationPaged(
-            currentPersona?.identifier,
-            {
-                network: 'all',
-                pageOffset: 0,
-            },
-            1000,
-        )
-        if (values.length === 0) return EMPTY_LIST
-        const friends: Friend[] = (
-            await Promise.all<Promise<Friend | undefined>>(
-                values.map(async (x) => {
-                    if (isProfileIdentifier(x.profile)) {
-                        const res = first(await Services.Identity.queryProfilesInformation([x.profile]))
-                        if (res?.linkedPersona !== undefined && res?.linkedPersona !== currentPersona?.identifier)
-                            return {
-                                persona: res.linkedPersona,
-                                profile: x.profile,
-                                avatar: res.avatar,
-                            }
-                        return
-                    } else {
-                        if (x.profile !== currentPersona?.identifier) return { persona: x.profile }
-                        return
-                    }
-                }),
-            )
-        ).filter(emptyFilter)
-        const allSettled = await Promise.allSettled(
-            friends.map((item) => {
-                const id = item.persona.publicKeyAsHex
-                return NextIDProof.queryProfilesByPublicKey(id, 2)
-            }),
-        )
-        const profiles: FriendsInformation[] = allSettled.map((item, index) => {
-            if (item.status === 'rejected') {
-                if (friends[index].profile) {
-                    return {
-                        profiles: [
-                            {
-                                platform: NextIDPlatform.Twitter,
-                                identity: friends[index].profile!.userId,
-                                is_valid: true,
-                                last_checked_at: '',
-                                name: friends[index].profile!.userId,
-                                created_at: '',
-                            },
-                        ],
-                        ...friends[index],
-                        id: friends[index].persona.publicKeyAsHex,
-                    }
+    return useInfiniteQuery(
+        ['friends', currentPersona],
+        async ({ pageParam }) => {
+            if (!cachedValues.current.length && currentPersona) {
+                const values = await Services.Identity.queryRelationPaged(
+                    currentPersona?.identifier,
+                    {
+                        network: 'all',
+                        pageOffset: 0,
+                    },
+                    3000,
+                )
+                if (values.length === 0) return EMPTY_LIST
+                cachedValues.current = values
+            }
```

Comment:
You can split into multiple queries, and control the order by `enabled`

https://tanstack.com/query/v4/docs/react/guides/dependent-queries


```ts
    // Get records first
    const { data: records = EMPTY_LIST, isLoading } = useQuery(['relation-records'], async () => {
        return Services.Identity.queryRelationPaged(
            currentPersona?.identifier,
            {
                network: 'all',
                pageOffset: 0,
            },
            3000,
        )
    })
    const allProfiles = useMemo(() => records.map((x) => x.profile), [records])
    useInfiniteQuery({
        enabled: !isLoading,
        queryKey: ['friends', currentPersona?.identifier.rawPublicKey, allProfiles],
        queryFn: async ({ pageParam }) => {
            const offset = pageParam || 0
            const profile = allProfiles[offset]
            if (isProfileIdentifier(profile)) {
                const friend = first(await Services.Identity.queryProfilesInformation([profile]))
                return {
                    friend,
                    nextOffset: offset + 1,
                }
            } else {
                return {
                    friend: { persona: profile },
                    nextOffset: offset + 1,
                }
            }
            // You an query multiple profile each page.
            // const chunk = 10
            // const profiles = allProfiles.slice(offset, offset + chunk)
            // const promises = profiles.map(x => { GET_PROFILES_BY_CHUNK })
            // return {
            //   friends: await promises
            //   nextOffset: nextOffset + chunk
            // }
        },
        getNextPageParam: (lastPage) => {
            if (lastPage?.nextOffset >= allProfiles.length) return
            return lastPage?.nextOffset
        },
    })

```